### PR TITLE
[vLLM] Don't force fp32_dest_acc_en=False graph-wide in benchmark configs

### LINKED ISSUE #5116: [vLLM] Per-user last-token gather index computed via bf16 matmul → wrong logits for batched users past flat-index 512
## Problem

Greedy (`temperature=0`) decoding of **N identical prompts in one batch** does not produce identical outputs across all rows. Each row in a batch is an independent computation, so argmax decoding must be identical for every row — divergence is a correctness bug. First spotted in `tests/benchmark/test_vllm_benchmarks.py::test_vllm_benchmark[llama-3.2-3b]` locally and on Nightly CI (https://github.com/tenstorrent/tt-xla/actions/runs/27175829468/job/80224459033#step:15:123), reproducible using simpler case below.

The divergence has a clean, structural signature: the batch splits into **contiguous groups of rows** that each compute a different result (e.g. rows `0..k-1` emit one token, rows `k..N-1` emit another). It is **deterministic**, and it is **fully eliminated by enabling fp32 destination accumulation** (`fp32_dest_acc_en=True`). It is therefore a **bf16 destination-accumulation** problem in the batched prefill graph (tt-mlir compute-kernel-config family).

## Reproduction

Llama-3.2-3B-Instruct, 32 identical prompts, greedy, generating a single token (pure prefill):

```python
import vllm

PROMPT = "Here is an exhaustive list of the best practices for writing clean code:"
N = 32

llm = vllm.LLM(
    model="meta-llama/Llama-3.2-3B-Instruct",
    max_model_len=128,
    max_num_seqs=N,
    max_num_batched_tokens=N * 128,        # prefill all N prompts together
    gpu_memory_utilization=0.5,
    enable_prefix_caching=False,
    additional_config={"fp32_dest_acc_en": False},
)
sp = vllm.SamplingParams(temperature=0.0, max_tokens=1, ignore_eos=True)
outs = llm.generate([PROMPT] * N, sp)

ids = [o.outputs[0].token_ids[0] for o in outs]
for i, t in enumerate(ids):
    print(f"row {i:2d}: token {t}")
print("unique tokens:", len(set(ids)))   # bug: 2 (should be 1)
```

Observed (deterministic): rows 0–3 emit token `720`, rows 4–31 emit token `320`; `unique tokens = 2`. With `fp32_dest_acc_en=True` all 32 rows emit `720` (`unique tokens = 1`). Full captured output: https://gist.github.com/kmabeeTT/dd7524de473cd8158040fc860886cafd

Reproduced at tt-xla `8a54156b5` (`main`) with stock tt-mlir `c5f398432` (the version `main` pins), single device.

## Observations (to guide debugging)

1. **Contiguous-group split, boundary tracks the prefill shape.** The batch splits into a contiguous prefix `0..k-1` vs the rest `k..N-1`. Where the boundary `k` falls depends on the prefill shape (the program-config / core-grid the batched ops are given), **not** on a fixed user count:

   | additional_config | first divergent row |
   |:--|:--|
   | `{fp32_dest_acc_en: False}` | **4** (group 0–3 vs 4–31) |
   | `{fp32_dest_acc_en: False, min_context_len: 32}` | **16** (group 0–15 vs 16–31) |

   Forcing a 32-token minimum prefill context moves the boundary from 4 to 16. This is consistent with the cause being the **batch grouping of the runtime-selected program-config**, not a magic number. (Batches at or below the boundary are clean: with the boundary at 16, batch 4/8/16 all agree and batch 24 splits 16/8.)

2. **Deterministic.** Re-running an identical config gives byte-identical per-row output every time (same split, same token ids). Not a race.

3. **Layer-count independent.** Same behavior on a single-decoder-layer truncation (`additional_config={"num_hidden_layers": 1}`) and on the full 28-layer model; only the specific token ids differ.

4. **Not the sampler.** Forcing host-side sampling (`additional_config={"cpu_sampling": True}`) reproduces the split, so the wrong values are already in the **logits coming off the device**, not in on-device sampling/argmax.

5. **`fp32_dest_acc_en=True` eliminates it** — all rows agree. So the divergence is entirely a low-precision (bf16) destination-accumulation effect. (Note: enabling it on 8B-class models has been associated with a separate DRAM-OOM regression, so it is not a free global fix.)

6. **No single op reproduces it in isolation — even config-matched.** A standalone `ttnn` program running each production prefill op (the QKV / O / gate-up / down matmuls, dense causal SDPA, rms_norm) at the exact batched shapes with **N identical rows** produces **bit-identical** outputs across all rows — including when passed the *exact* compute config the lowered graph specifies (`math_fidelity=hifi4, fp32_dest_acc_en=False`). A row-local op (matmul / elementwise / norm) cannot, by construction, turn identical input rows into a grouped split. So the split is **introduced by graph composition**, not by any single kernel.

7. **Layouts are not the asymmetry.** All intermediate tensors in the lowered prefill graph are **interleaved DRAM** (no sharded layouts), and the matmul ops carry only `compute_config = hifi4` with **no explicit program_config** — the core grid is selected from the activation shape at runtime.

## Likely root cause

The combination of (5)+(6)+(7) points at the **in-graph program-config / core-grid that tt-metal selects for the batched prefill ops under the compiler-emitted `hifi4, fp32_dest_acc_en=False` config**: it partitions the batch into contiguous groups whose bf16 destination-accumulation reductions round differently, so one group's logits diverge enough to flip the greedy token. The boundary moving with the prefill shape (4 vs 16 above) is direct evidence of this grouping. fp32 destination accumulation has enough headroom that the groups agree.

This is the same family as the known compute-kernel-config issues:
- [tt-mlir #8666](https://github.com/tenstorrent/tt-mlir/issues/8666) — `fp32_dest_acc_en` not applied properly for `ttnn.matmul`.
- [tt-mlir #8657](https://github.com/tenstorrent/tt-mlir/issues/8657) — SDPA has no compute-kernel-config / fp32-accumulation path (`ttnn.scaled_dot_product_attention*` don't implement `TTNNComputeKernelConfigOpInterface`).

## Suggested next step

Localizing the exact op needs an **in-graph per-op activation dump** (tt-mlir runtime op callback / `dumpTensor`): run the batched prefill, dump every op's output, and find the first op whose output for a row past the boundary diverges from row 0 for identical inputs. That op's emitted program-config / core-grid is the one introducing the grouping asymmetry.

## Also visible in the existing benchmark

The same divergence surfaces in the existing `tests/benchmark/test_vllm_benchmarks.py -k llama-3.2-3b` run (batch 32, `fp32_dest_acc_en=False`), which prints multiple distinct completions for 32 identical prompts but still passes. That path generates 128 tokens, so it shows the divergence in a noisier form (decode mixed in); use the single-token snippet above to see the clean structural signature.

It is already live in nightly CI — e.g. the `perf vllm_llama_3_2_3b (n150-perf)` job at https://github.com/tenstorrent/tt-xla/actions/runs/27064401476/job/79883258025#step:14:121 (2026-06-06) printed 3 distinct completions for the 32 identical prompts and was still marked **success**.

## Why it isn't caught today

Benchmarks assert token counts and PCC of **User 0** only, so per-user divergence (Users 1..N) is invisible — a batch of 32 identical prompts can print 2 distinct completions and still pass.

### PR DESCRIPTION
**Pull request by dgolubovicTT**
### Ticket
Closes #5116

### Problem description
TLDR: Inside `tests/benchmark/test_vllm_benchmarks.py ` we were setting `fp32_dest_acc_en` to False by default which caused divergence in batched prefil outputs. Setting this knob inside benchmark config is wrong. Here is the logic that this knob follows inside mlir compiler part when it is not explicitly set by frontend:
- For opt level 0: we don't care about perf and default value of `fp32_dest_acc_en` is True.
- For opt level >0: we care about perf and compiler leaves `fp32_dest_acc_en` unset, so each ttnn op can choose its value based on inputs etc.
Same is true for math_fidelity knob. For opt level 0 its default is HiFi4, but when opt level is >0, we leave it unset because of the perf.

Here is more detailed explanation why setting `fp32_dest_acc_en` to False globally caused per batch output mismatch:

Batched greedy decoding of N identical prompts diverged into contiguous groups: users past flat-index 512 emitted a different token than users before it. Root cause is a graph-wide fp32_dest_acc_en=False override that the benchmark _config passed by default. TTNNSetComputeKernelConfig stamps that value onto every matmul exposing a compute-kernel-config, including the tiny fp32 index matmul that lowers the per-user last-token gather (flat_index = [batch_idx, pos] @ [stride, 1]). With bf16 destination accumulation, flat indices >= 512 round to a multiple of 4, so the wrong hidden-state row is gathered -> wrong logits -> divergent argmax token. At opt>0 the pipeline resets the override to unset (so it was masked in nightly); at opt=0 the override survives and it breaks.

A compute-kernel-config knob should only ever be set deliberately, never carried as a frontend default. Leaving fp32_dest_acc_en unset lets ttnn pick the per-op default (fp32 accumulation for fp32-output ops like the index matmul, bf16 for bf16 compute matmuls) -- exact index arithmetic with no memory regression on the big matmuls.

### What's changed

Removing default False value from vllm benchmarks and old llm benchmarks.
- tests/benchmark/test_vllm_benchmarks.py: _config default False -> None
- tests/benchmark/test_llms.py: drop hardcoded False on the two 8B configs
Verified (Llama-3.2-3B, 32 identical prompts, greedy, opt0): before the change all rows split (boundary row 5, tokens [320, 4815]); after, all 32 rows agree (720). Deliberate True/False overrides are still honored.

Also cleaned up two vLLM prefill integration tests that carried the same fp32_dest_acc_en=False default, and added a comment so it doesn't get reintroduced:

- tests/integrations/vllm_plugin/generative/test_b1_prefill.py and test_chunked_prefill.py: drop the fp32_dest_acc_en=False from their configs. Both run at opt level 1, where the compiler already resets the knob to unset — so the value was a dead no-op (not a functional change), just the same anti-pattern. Re-ran both without it and nothing changed: test_chunked_prefill.py 3/3 passed, test_b1_prefill.py passed with identical TTFT behavior.

### CODE REVIEW

**Review by mmanzoorTT — APPROVED** [review]

File: integrations/vllm_plugin/vllm_tt/platform.py [hunk 1]
```diff
@@ -128,6 +128,17 @@ class TTConfig:
...
+    # PREFER LEAVING THIS None (unset). Do NOT set it to False as a default /
+    # convenience in a model config or test. It is a GRAPH-WIDE override applied
+    # to every matmul, including tiny index-arithmetic matmuls (e.g. the per-user
+    # last-token gather that lowers to `flat_index = indices @ strides` ->
+    # embedding). Forcing bf16 destination accumulation there rounds flat
+    # indices >= 512 to the wrong value -> wrong gathered row -> wrong logits ->
+    # per-user divergence in batched greedy decoding (tt-xla #5116). Unset is
+    # better for accuracy: ttnn picks fp32 accumulation for fp32-output ops
+    # (exact index math) and bf16 for bf16 compute matmuls (no memory
+    # regression). Only set True/False deliberately for a validated reason.
```
Comment (jameszianxuTT) [inline integrations/vllm_plugin/vllm_tt/platform.py:141]: Great find! 

Not to block now, but I think this is a footgun that should not be covered up by a comment.  Maybe even just a rename to something like `override_all_ops_fp32_dest_acc_en` to make someone question their usage of this, since from userland we use this option without much thought.

**Review by jameszianxuTT — APPROVED** [review]

File: integrations/vllm_plugin/vllm_tt/platform.py (hunk 1 above)
Comment (jameszianxuTT) [reply integrations/vllm_plugin/vllm_tt/platform.py:141]: Or removing user ability to override this from compiler_options directly, as it seems that setting graph level fp32 accumulation to either TRUE/FALSE has little utility and can introduce confounding errors. idk if we have some mechanism to enforce this on op level, but maybe leaving this kind of feature up to dedicated tooling with a different integration point instead of as a first class compile option may be a good idea, though my context here is limited

**Review by bmalesevicTT — APPROVED** [review]

**Review by vkovacevicTT — APPROVED** [review]

**Review by mmanzoorTT — APPROVED** [review]

File: integrations/vllm_plugin/vllm_tt/platform.py (hunk 1 above)
Comment (dgolubovicTT) [reply integrations/vllm_plugin/vllm_tt/platform.py:141]: Thanks, I will think about this. I got a request from training team to expose all other compute kernel cfg knobs because they need them for additional accuracy (for example math approx mode). So I will tie everything in next PR related to this...
